core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -670,6 +758,21 @@ private synchronized SplitAssignment getSplitAssignment(PlanNodeId planNodeId)
         boolean pendingNoMoreSplits = Boolean.TRUE.equals(this.noMoreSplits.get(planNodeId));
         boolean noMoreSplits = this.noMoreSplits.containsKey(planNodeId);
 
+        // limit the number of splits only for a partitioned source
+        if (planFragment.getPartitionedSourceNodes().contains(planNodeId) && splitBatchSize < splits.size()) {
+            if (splitBatchSizeChanged) {
+                log.info("%s - Splits are limited by splitBatchSize: splitBatchSize=%s, splits=%s, planNodeId=%s", taskId, splitBatchSize, splits.size(), planNodeId);
+                // set this false to prevent repetitive logs
+                splitBatchSizeChanged = false;
+            }
+            splits = splits.stream()
```

Comment:
Maybe use of `linkedHashSet` would be a better approach here?

Reply:
My intention was to avoid sorting splits by having splits already inserted into a set which is sorted by sequenceId and just do poll of first splits from this set.

Reply:
I don't understand what you meant. (especially `just do poll of first splits from this set`) AFAIK, `splits` does not guarantee the `sequenceId` order.